### PR TITLE
Support for `@threads @simd for`

### LINKED ISSUE #32684: Cannot combine @simd and @threads on a loop
```julia
julia> import Base: Threads

julia> @simd Threads.@threads for i in 1:10 end
ERROR: LoadError: Base.SimdLoop.SimdError("for loop expected")
Stacktrace:
 [1] compile(::Expr, ::Bool) at ./simdloop.jl:54
 [2] @simd(::LineNumberNode, ::Module, ::Any) at ./simdloop.jl:124
in expression starting at REPL[2]:1

julia> Threads.@threads @simd for i in 1:10 end
ERROR: LoadError: ArgumentError: unrecognized argument to @threads
Stacktrace:
 [1] @threads(::LineNumberNode, ::Module, ::Vararg{Any,N} where N) at ./threadingconstructs.jl:101
in expression starting at REPL[4]:1
```

### PR DESCRIPTION
**Pull request by ghost**
Partly fixes #32684. By this change, `@threads @simd for` seems to work, but not yet for `@simd @threads for`. I don't know what the expected behavior of `@simd @threads for` is.  It would be helpful if someone had any advice.

### CODE REVIEW

File: base/threadingconstructs.jl [hunk 1]
```diff
@@ -56,9 +56,16 @@ function _threadsfor(iter,lbody)
                 end
             end
             # run this thread's iterations
-            for i = f:l
-                local $(esc(lidx)) = @inbounds r[i]
-                $(esc(lbody))
+            if $simd == true
+                @simd for i = f:l
```
Comment (vchuravy) [inline base/threadingconstructs.jl]: This misses handling for `@simd ivdep`.

**Review by vchuravy — COMMENTED** [review]
Generally speaking I think we don't need to worry about `@threads @simd` since `@simd` is really impactful if there is a reduction operation in the loop (in most other cases LLVM will figure it out on its own). But if you are multi-threading a reduction naively you will introduce a race and I think if you do it correctly `@simd` won't help much. So it is almost always better to write the reduction as a nested for-loop over the blocked region and only `@simd` the inner loop.

**Comment by mikan2087** [comment]
Thanks for the comment. I understood your explanation, and now I think I don't need to do anything to support `@simd @threads`, which returns ERROR.

And I committed a fix for `@threads @simd ivdep`.

**Comment by vchuravy** [comment]
I still don't quite understand the use-case for `@threads @simd`. Is there are particular case where you have found it useful? Also needs a test case. Another thought was that since you already have a linear index inside the threads loop you could just emit the `Expr(:loopinfo, ... ` directly instead of using the simd macro again.

**Comment by vchuravy** [comment]
@mikan2087 my last comment sounds a bit harsh in retrospect. Thank you for all the work you have put into this sofar!

While I am questioning the usefulness of the feature it doesn't mean we shouldn't support it, since it will fix a bug.
It would be fantastic if you could add a test in https://github.com/JuliaLang/julia/blob/master/test/threads_exec.jl

**Comment by KristofferC** [comment]
If there is no example where `@simd` on a threaded loop leads to faster execution then just error with a better error message would be preferable, no?

**Comment by DilumAluthge** [comment]
It looks like this pull request is no longer mergeable.

If you would still like to make this change, please make a new pull request based on the latest `master` branch of Julia.

**Comment by xwuupb** [comment]
> I still don't quite understand the use-case for `@threads @simd`. Is there are particular case where you have found it useful? ...

SAXPY, matrix-vector multiplication, matrix-matrix multiplication, etc I know I can use BLAS. But `@threads @simd` is **critically important** for (customized) linear algebra operations in user applications, because both the thread-level parallelization and instruction-level parallelization are essential to achieve the desire performance.

**Comment by timholy** [comment]
https://github.com/JuliaSIMD/LoopVectorization.jl

**Comment by xwuupb** [comment]
> https://github.com/JuliaSIMD/LoopVectorization.jl

I was expecting such important functionality for parallelization as well as code performance is supported by Julia without external packages.

**Pull request closed without merging**